### DISCUSSION TITLE
Journey Collection

### DISCUSSION THREAD

**Discussion by p-mathias** [odata2ts/odata2ts#210] [Ideas]
Since many thing happen to get odata2ts running,
i would like to collection this experience,
with the hope that some derive can be make based on this experience:
Derive as checkmark fields

i didn't understand the emit-modes: https://odata2ts.github.io/docs/generator/configuration#emit-modes
i had to change it to ts,
which suprised me in general since the lib called odata2TS and not odata2JS/DTS

- [ ] is the default mode of emit-modes the best?

the default recommandation to generate the builded ts files in a build directory is uncommon in my eyperience,
for me it's more common to generate the this kind information into node_modules directory like prisma it does:
https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/generating-prisma-client#the-prismaclient-npm-package

- [ ] revist the defaule/recommaned place to safe the generate ts files

in https://odata2ts.github.io/docs/odata/odata-types is not referenz that a ulti function exists which could help to handle the data

- [ ] ref from odata-types to ulti 

duplicate between the section generator, query builder, client like query building in query builder and client section, etc.
maybe a legacy thing because everything is/was equal, but my feeling say, it would be better the doc build up from the feature riches product (your client),
and below than describe the generator and query build, so that query buiding needed to defined once in a sub chapter of query-builder which is a sub chapter of client,
or work with with more cross referenz (would it make harder to read)

- [ ] redurce duplicate

about provided api, from a odata pov make it sense, but it's drive me crazy,
that set function can provide a collection or single entity,
which make it also harder to understand the ts information.
so from my external view i would like to had the set function but behavior that they return one entity and
secondly a serverAllSet (what ever the name is) that return a collection.
i mean i could do this by hand,
but from you yt video from ui5.com i understand that you target dev is not a odata expert,
so when this is the target user group sutch abstraction would help

- [x] a far generalize odata abstraction

something what i doesn't understand why date information get to me in this format that they are,
i understand that v2 and v4 does it differently, but why should i care?
end of the day i got a v2 date and i used this code (since the information that ulti lib exists what not mention in the doc part where i looking): ``new Date(Number.parseInt(ele.begda.substring(6, 19)))``

- [x] provide methode how are generalize across v2/v4 in js focus way -> most liky that i want a Date object, which is delivered by the generator per data field

i had to setup local proxy for cors stuff
since i already use vite for bundling is also use his proxy
vite.config.ts:
```
    server: {
      proxy: {
        '/endPoint': { // localhost:5173/endPoint
          target: 'https://company_server.com/path_to_endPoint', // https://abap_server.com/sap/opu/odata/sap/
          changeOrigin: true,
          secure: false,

          configure: (proxy, options) => {
            const username = process.env.VITE_ABAP_USER
            const password = process.env.VITE_ABAP_PASSWORD
            options.auth = `${username}:${password}`
          }
        }
      }
    },
```

in .env i define my abap username and pw
```
VITE_ABAP_USER="USERNAME"
VITE_ABAP_PASSWORD="Secret"
```
to get env in vite config
loadEnv needs to be imported
process.env needs to be loaded: ``   process.env = { ...process.env, ...loadEnv(mode, process.cwd()) }``

since i didn't like to had the url end point in my code i also use the odata2ts.config.ts to read the url to my service:
```
  services: {
    endPoint: {
      sourceUrl: import.meta.env.PROD
        ? 'https://abap_server.com/sap/opu/odata/sap/endPoint'
        : 'http://localhost:5173/endPoint',
      source: 'resource/endPoint.xml',
      output: 'build/endPoint'
    },
```
so when i run in local dev the proxy is taken on build/preview the real server endPoint is used.

- [x] since that some dev came from sapui5 env, where everything is layout for them, you had at least mention that something is needed and referenz maybe some source to setup sutch local env.

- [ ] align doc between odata2ts config between client and generator about handle sourceUrl [generator](https://odata2ts.github.io/docs/generator/setup-and-usage#config-file) [client](https://odata2ts.github.io/docs/odata-client/overview-and-setup#initialization)

about adding parameter to the url:
i created a own fetch class and extend the get/put methode:
```ts
class myFetch extends FetchClient {
  constructor(...args) {
    super(args)
  } 
  public post(url, data, requestConfig, additionalHeaders) {
    url = url + '?sap-client=200'
    return super.post(url, data, requestConfig, additionalHeaders)
  }

  public get(url, requestConfig, additionalHeaders) {
    url = url + '?sap-client=200'
    return super.get(url, requestConfig, additionalHeaders)
  }
}
```

issue: post doesn't work, since fall in csrf trap, which i try to solve by the mention doc way https://odata2ts.github.io/docs/odata-client/http-client/fetch#automatic-csrf-token-handling

- [ ] Extend the docs with information (a url to mdn or so) what csrf is 

but hard to see where stuf get wrong here
this my journey atm

**Comment by texttechne** [comment]
Hi @p-mathias,
thanks for taking the time to document your journey. You bring up some really good points and I need some time to think about them.

But a couple of things we can clarify right now.

### CORS
This is a typical front end development issue and using a proxy to solve that issue is common place and IMO the best way to tackle the issue. But this has nothing to do with odata2ts per se. However, I could add some pointers in the getting started guide for the OData client.

### Data Types
I think you've missed some documentation here, since you've overlooked the topic of **converters**.
See [Types and Converters](https://odata2ts.github.io/docs/generator/configuration#types-and-converters). 
To get a JS `Date` you would use the `common-converter` package. If you're using a V2 service you would additionally add the `v2-to-v4-converter`.

Maybe an additional cross reference section in the getting started guide would help. Something like additional topics to look into...

### CSRF Handling
Did you have any trouble here? 

The additional link in the section is a good idea.

### Provided Service API
It took me some time to get to this point, but I actually love the API and it plainly feels right:
```
trippinService.people()
trippinService.people("russelwhyte")
```
It reflects how OData works, is a succinct notation and prevents any name clashes, which might happen if you introduce additional custom methods in the service class.

So I have to disappoint you there... it will stay the way it is.

---

So much for now. As I said, I need some time to thinks about the other stuff...

**Comment by p-mathias** [comment]
> ### Data Types
> I think you've missed some documentation here, since you've overlooked the topic of **converters**. See [Types and Converters](https://odata2ts.github.io/docs/generator/configuration#types-and-converters). To get a JS `Date` you would use the `common-converter` package. If you're using a V2 service you would additionally add the `v2-to-v4-converter`.
> 
> Maybe an additional cross reference section in the getting started guide would help. Something like additional topics to look into...

in my eyes less places where the same/similiar information is placed would be better,
i looked for exp. here for sutch information: https://odata2ts.github.io/docs/odata/odata-types and didn't found it
 
> ### CSRF Handling
> Did you have any trouble here?
> 
> The additional link in the section is a good idea.

yes still got problem,
but last week i hadn't time to look deeper into this.

**Comment by p-mathias** [comment]
about information about csrf, this blog post is maybe a good idea: https://blogs.sap.com/2021/06/04/how-does-csrf-token-work-sap-gateway/
to use it in the documenation
